### PR TITLE
Tika: Handle topic name conflict

### PR DESCRIPTION
**Pull request by SEQUOIIA**
Story details: https://app.clubhouse.io/devex/story/1069

### CODE REVIEW

**Comment by kim-lindhard-dfds** [comment]
[//]: # (4744e426-5795-11e9-8647-d663bd873d93)

|| **CodeScene Delta Analysis Results** |
| --- | ---- |
| **Risk** |10|
| **Quality Gates** |OK|
| **Description** |The change is high risk and adds 148 lines, deletes 3 lines of code in 8 files. The risk is lower since it is a very experienced author with many contributions.|
| **Warnings** |Delta Analysis Info<br/><ul><li>3 files in this change set did not exist with the same name on the main branch. This is probably because those files have been renamemd. Skipping code health checks for those files.</li></ul>|

File: server/src/server/wrapper/connected/CCloudTopics.ts [hunk 1]
```diff
@@ -15,12 +17,19 @@ export class CcloudTopics implements Topics {
...
-        await executeCli([
-            "kafka", "topic",
-            "create", name,
-            "--partitions", partitionCount + "",
-            "--cluster", process.env.TIKA_CCLOUD_CLUSTER_ID
-        ]);
+        let topics = await this.getTopics();
+
+        let topicFound = topics.find(v => v.valueOf() === name.valueOf());
+        if (topicFound === undefined) {
```
Comment (kim-lindhard-dfds) [inline server/src/server/wrapper/connected/CCloudTopics.ts:23]: Could we save a call by trying to create the topic and then throwing if the response states that the topic name exits?

File: server/src/server/wrapper/model/error.ts [hunk 2]
```diff
@@ -5,3 +5,11 @@ export class CcloudSessionExpiredException extends Error {
     this.message = "Login session has expired";
   }
 }
+
+export class CcloudTopicAlreadyExistsException extends Error  {
```
Comment (kim-lindhard-dfds) [inline server/src/server/wrapper/model/error.ts]: Do we need to indicate that the error stems from ccloud?

File: server/src/server/api/TopicsInterface.ts [hunk 3]
```diff
@@ -12,13 +14,29 @@ export class TopicsInterface {
...
-            );
-
-            res.sendStatus(200);
+            let good : boolean = true;
+
+            try {
+                await topic.createTopic(
+                    req.body.name,
+                    parseInt(req.body.partitionCount),
+                );
```
Comment (kim-lindhard-dfds) [inline server/src/server/api/TopicsInterface.ts:23]: if you put the ` res.sendStatus(200);` here then you can throw away the good variable

File: clients/dotnet-core-rest/src/Tika.RestClient/Features/Topics/TopicsClient.cs [hunk 4]
```diff
@@ -39,10 +42,20 @@ public async Task CreateAsync(TopicCreate topicCreate)
...
-            await _httpClient.PostAsync(
+            var res = await _httpClient.PostAsync(
                 new Uri(TOPICS_ROUTE, UriKind.Relative),
                 content
             );
+
+            if (res.StatusCode == HttpStatusCode.Conflict)
+            {
+                var rawText = await res.Content.ReadAsStringAsync();
+                var responseData = JsonConvert.DeserializeObject<TikaGenericError>(rawText);
```
Comment (kim-lindhard-dfds) [inline clients/dotnet-core-rest/src/Tika.RestClient/Features/Topics/TopicsClient.cs:53]: [np] should responseData be error or errorResponse?

File: clients/dotnet-core-rest/src/Tika.RestClient.IntegrationTests/Features/Topics/SameTopicNameTwice.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                    name = _topicName,
+                    partitionCount = 1
+                };
+
+                await _client.Topics.CreateAsync(topicCreate);
+            }
+            catch (TopicAlreadyExistsException e)
+            {
+                _topicAlreadyExistsException = e;
+                Console.WriteLine(e);
```
Comment (kim-lindhard-dfds) [inline clients/dotnet-core-rest/src/Tika.RestClient.IntegrationTests/Features/Topics/SameTopicNameTwice.cs:58]: [np] do we need it?

**Review by kim-lindhard-dfds — APPROVED** [review]
[np] == nitpick

File: server/src/server/wrapper/connected/CCloudTopics.ts (hunk 1 above)
Comment (SEQUOIIA) [reply server/src/server/wrapper/connected/CCloudTopics.ts:23]: Sure. It's a fair bit more work and some extra parsing, but that could be done.

**Comment by kim-lindhard-dfds** [comment]
[//]: # (4744e426-5795-11e9-8647-d663bd873d93)

|| **CodeScene Delta Analysis Results** |
| --- | ---- |
| **Risk** |10|
| **Quality Gates** |OK|
| **Description** |The change is high risk and adds 155 lines, deletes 10 lines of code in 8 files. The risk is lower since it is a very experienced author with many contributions.|
| **Warnings** |Delta Analysis Info<br/><ul><li>3 files in this change set did not exist with the same name on the main branch. This is probably because those files have been renamemd. Skipping code health checks for those files.</li></ul>|

File: server/src/server/wrapper/model/error.ts (hunk 2 above)
Comment (SEQUOIIA) [reply server/src/server/wrapper/model/error.ts]: We don't _need_ to. 275a6bb0

**Pull request merged**